app/models/order.rb

```diff
@@ -128,6 +136,20 @@ def delete_orders_packages
     end
   end
 
+  def international_orders?
+    [Type::SHIPMENT, Type::CARRY_OUT].include?(detail_type)
+  end
+
+  def self.get_subsequent_international_code(params)
+    record = Order.where(detail_type: params).order("id desc").pluck(:code)
+    unless record.empty?
+      susequent_value = record.map{|s| s.gsub(/\D/, "")}
+      susequent_value.map!(&:to_i).max + 1
+    else
+      1000
+    end
+  end
```

Comment:
```suggestion
# missing_code will give the next code
# ex: if 0001 and 0004 exists, missing_code will give 2, which will be converted to 0002
# If there is no missing_code, then next_code will give 0005
def self.get_next_code_for(type)
    code = missing_code_for(type) || next_code(type) + 1
    code.to_s.rjust(5, '0')
  end

  def self.missing_code_for(type)
    query = <<-query
      select s.i as first_missing_code from generate_series(1, :max) s(i) WHERE s.i not in (
          select CAST(substring(orders.code, 2) as INTEGER) as code from orders where orders.detail_type = :type and substring(orders.code, 2) ~ '^\\d+$'
      ) limit 1
    query

    code = connection.exec_query(sanitize_sql_array([query, type: type, max: Order.where(detail_type: type).count])).first['first_missing_code']
    code
  end

  def self.next_code(type)
    query = <<-query
    SELECT max(CAST(substring(orders.code, 2) as INTEGER)) as code from orders where orders.detail_type = :type and substring(orders.code, 2) ~ '^\d+$'
    query

    code = connection.exec_query(sanitize_sql_array([query, type: type])).first["code"]
    code
  end
```

Reply:
You can refactor the above
It has repetitive queries, you can extract them to separate method if possible

Also you can move to a separate concern or lib : )

Reply:
we probably want to consolidate the generation of shipment codes with the one of goodcity orders

There's an existing `generate_gc_code` method. We should not merge this PR having 2 separate entry points for generating codes.

We can probably end with a single entry point, maybe like `generate_code(detail_type)`

Reply:
There is a toggle of Shipment and Carryout in the shipment form so thought of doing it with different entry, will do it with generate_code(detail_type)